akka-typed/src/main/scala/akka/typed/internal/EventStreamImpl.scala

```diff
@@ -0,0 +1,332 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com/>
+ */
+package akka.typed
+package internal
+
+import akka.{ actor ⇒ a, event ⇒ e }
+import java.util.concurrent.atomic.AtomicReference
+import scala.annotation.tailrec
+import scala.concurrent.{ Await, Promise }
+import akka.util.{ ReentrantGuard, Subclassification, SubclassifiedIndex }
+import scala.collection.immutable
+import java.util.concurrent.TimeoutException
+import akka.util.Timeout
+
+/**
+ * An Akka EventStream is a pub-sub stream of events both system and user generated,
+ * where subscribers are ActorRefs and the channels are Classes and Events are any java.lang.Object.
+ * EventStreams employ SubchannelClassification, which means that if you listen to a Class,
+ * you'll receive any message that is of that type or a subtype.
+ *
+ * The debug flag in the constructor toggles if operations on this EventStream should also be published
+ * as Debug-Events
+ */
+private[typed] class EventStreamImpl(private val debug: Boolean)(implicit private val timeout: Timeout) extends EventStream {
+  import e.Logging._
+  import ScalaDSL._
+  import EventStreamImpl._
+
+  private val unsubscriberPromise = Promise[ActorRef[Command]]
+  private val unsubscriber = ActorRef(unsubscriberPromise.future)
```

Comment:
I really like that feature btw. Was talking about it so many times with various people, so good to finally have it :-)